src/librustc_middle/ty/relate.rs

```diff
@@ -510,12 +510,21 @@ pub fn super_relate_consts<R: TypeRelation<'tcx>>(
     let tcx = relation.tcx();
 
     let eagerly_eval = |x: &'tcx ty::Const<'tcx>| {
+        // FIXME(eddyb) this doesn't account for lifetime inference variables
+        // being erased by `eval`, *nor* for the polymorphic aspect of `eval`.
+        // That is, we could always use `eval` and it will just return the
+        // old value back if it doesn't succeed.
         if !x.val.needs_infer() {
             return x.eval(tcx, relation.param_env()).val;
         }
         x.val
     };
 
+    // FIXME(eddyb) doesn't look like everything below checks that `a.ty == b.ty`.
+    // We could probably always assert it early, as `const` generic parameters
+    // are not allowed to depend on other generic parameters, i.e. are concrete.
```

Comment:
I think this temporary and won't always remain so, so let's not bake that assumption too deeply into the compiler.

Reply:
We have already, this is nothing compared to everything else. It's only "temporary" in the sense that the first thing we stabilize won't have generic parameters in `const` parameter types.

Reply:
Part of the problem is we'd need to stabilize something like `StructuralMatch` (some *deep* trait, unlike `StructuralPartialEq`/`StructuralEq`), because we can't just allow unrestricted types.
But even with that also it's a PITA to implement correctly, far more so than just concrete types.

Reply:
Sure I understand all of this; I'm just saying we should avoid too deep of an architectural dependence on the fact that const params can't depend on type params so that it doesn't take a year to dig out of.